### PR TITLE
FS-880 & FS-864 section support

### PR DESCRIPTION
**Pull request by harryyo**
Refactor to support new section put, status and application minting endpoints for save and return

### CODE REVIEW

File: database/store.py [hunk 1]
```diff
@@ -26,45 +27,196 @@ def applications_index(self) -> dict:
                 "id": application.get("id"),
                 "account_id": application.get("account_id"),
                 "status": application.get("status"),
+                "account_id": application.get("account_id"),
                 "fund_id": application.get("fund_id"),
                 "round_id": application.get("round_id"),
+                "project_name": application.get("project_name", ""),
+                "last_edited": application.get("last_edited"),
+                "started_at": application.get("started_at"),
```
Comment (SentientHousePlant) [inline database/store.py:35]: ⛏️ Can this not just be initialised here?

File: database/store.py [hunk 2]
```diff
@@ -26,45 +27,196 @@ def applications_index(self) -> dict:
...
+    def submit_application(self, application_id):
+        self._applications[application_id][
+            "date_submitted"
+        ] = datetime.datetime.now(datetime.timezone.utc).strftime(
+            "%Y-%m-%d %H:%M:%S"
+        )
+        self._update_statuses(application_id)
+        return self._applications[application_id]
+
+    def get_section(self, application_id, section_name):
```
Comment (SentientHousePlant) [inline database/store.py:72]: 🔧 Is it possible to get a docstring here?

File: database/store.py [hunk 3]
```diff
@@ -26,45 +27,196 @@ def applications_index(self) -> dict:
...
+        return self._applications[application_id]["sections"][section_name]
+
+    def _find_value_by_key(self, data: dict, target):
+        for key, value in data.items():
+            if isinstance(value, dict):
+                yield from self._find_value_by_key(value, target)
```
Comment (SentientHousePlant) [inline database/store.py:78]: 💯 ⛏️  This is really cool Harry, could we have some comments to explain your expert python skills? I really like the use of recursion and is instance!! <3

Comment (SentientHousePlant) [reply database/store.py:78]: 🤔 Why have we used yield instead of return :)

File: database/store.py [hunk 4]
```diff
@@ -26,45 +27,196 @@ def applications_index(self) -> dict:
...
+                # Update application statuses
+                self._update_statuses(application_id)
+        # Set last edited
+        self._applications[application_id][
+            "last_edited"
+        ] = datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S")
+        # IF section includes "your-project-name" field
+        # THEN update application project name
+        project_name = list(
+            self._find_value_by_key(new_json, "your-project-name")
```
Comment (SentientHousePlant) [inline database/store.py:116]: ⛏️ 🤔 Doesn't the use of list make the yield redundant?

File: tests/api_data/endpoint_data.json [hunk 5]
```diff
@@ -1,21 +1,51 @@
 {
```
Comment (SentientHousePlant) [inline tests/api_data/endpoint_data.json:1]: 📌 Would it be better to use monkey patches in the future instead of a test dictionary? (Or probably both)

**Review by SentientHousePlant — APPROVED** [review]
Just some small comments but overall very nice!!

File: database/store.py (hunk 1 above)
Comment (harryyo) [reply database/store.py:35]: started_at could be yes, last_edited currently presuming null to start with and then has a datetime when some answers have been added (rather than the form minted), and project_name is null to start with and then that get's added (currently) on the first page of the form.

File: api/namespace/applications/models/section.py [hunk 6]
```diff
@@ -0,0 +1,77 @@
+from api.namespace.applications.applications_ns import applications_ns
+from api.namespace.applications.models.metadata import metadata
+from api.namespace.applications.models.question import question
+from api.namespace.applications.models.question import question_status
+from flask_restx import fields
+
+section_inbound = applications_ns.model(
```
[the remark this answers is not in the document]
Comment (harryyo) [reply api/namespace/applications/models/section.py:7]: Thanks

File: api/namespace/applications/routes.py [hunk 7]
```diff
@@ -59,70 +63,87 @@ def get(self):
         }
         return APPLICATIONS.search_applications(args), 200, response_headers
 
+    create_application_parser = reqparse.RequestParser()
+    create_application_parser.add_argument("account_id", location="json")
+    create_application_parser.add_argument("round_id", location="json")
+    create_application_parser.add_argument("fund_id", location="json")
```
[the remark this answers is not in the document]
Comment (harryyo) [reply api/namespace/applications/routes.py:69]: yep, might be good to have an agreed convention for location for data depending on methods?

File: tests/api_data/endpoint_data.json (hunk 5 above)
Comment (harryyo) [reply tests/api_data/endpoint_data.json:1]: yes in short. I find the test dictionary is (or has been) handy for locally development in that it's just an easy to maintain version of api contracts (while the apis themselves are in development). But yes, should be more formally abstracted from code

File: database/store.py (hunk 3 above)
Comment (harryyo) [reply database/store.py:78]: I have to agree - totally pointless

File: database/store.py (hunk 4 above)
Comment (harryyo) [reply database/store.py:116]: yes

File: database/store.py (hunk 3 above)
Comment (harryyo) [reply database/store.py:78]: Would be useful if we expected there to be multiple keys with matching names (which there could be in other dictionaries) but here we're only expecting there to be one

File: api/namespace/applications/routes.py [hunk 8]
```diff
@@ -59,70 +63,87 @@ def get(self):
...
+        )
+        return application, 201
 
 
-@applications_ns.route("/<application_id>", methods=["GET", "POST"])
-class Application(Resource):
-    """
-    Operations on a single application
-    """
+@applications_ns.route("/sections")
```
Comment (endamccormack) [inline api/namespace/applications/routes.py:84]: why isnt this `/<application_id>/sections` ?

File: api/namespace/applications/routes.py [hunk 9]
```diff
@@ -59,70 +63,87 @@ def get(self):
...
+    put_section_parser.add_argument("name", location="json")
+
+    @applications_ns.doc(
+        "put_section", methods=["PUT"], parser=put_section_parser
+    )
+    @applications_ns.marshal_with(section, code=201)
+    def put(self):
+        request_json = request.get_json(force=True)
+        section_name = request_json["name"]
+        section_questions = request_json["questions"]
```
Comment (endamccormack) [inline api/namespace/applications/routes.py:97]: swagger doesn't seem to line up with this

File: api/namespace/applications/routes.py (hunk 8 above)
Comment (harryyo) [reply api/namespace/applications/routes.py]: Because (currently) afaik we can't dynamically update the callback url for individual PUT requests from the form runner, so we're getting the application_id from the metadata within the json in the request body

File: dockerfile [hunk 10]
```diff
@@ -0,0 +1,11 @@
...
+
+WORKDIR /app
+COPY requirements.txt requirements.txt
+RUN pip --no-cache-dir install --ignore-installed distlib -r requirements.txt
+RUN pip install gunicorn
+COPY . .
+
+EXPOSE 8080
+
+CMD ["gunicorn", "-w", "1", "-b", "0.0.0.0:8080", "app:create_app()"]
```
Comment (gidsg) [inline dockerfile:11]: This looks fine but in general I would rather unrelated things go into separate PRs rather than bundling in stuff which is not in the story.

File: api/namespace/applications/routes.py [hunk 11]
```diff
@@ -108,7 +108,10 @@ def put(self):
             application_id, section_name, section_dict
         )
 
-        return updated_section, 201
+        if updated_section:
+            return updated_section, 201
```
Comment (SentientHousePlant) [inline api/namespace/applications/routes.py:112]: ⛏️ Should this be "make_response"?

**Review by SentientHousePlant — APPROVED** [review]
lgtm ✨

File: dockerfile (hunk 10 above)
Comment (endamccormack) [reply dockerfile:11]: Good work but I agree, separate PR's causes less confusion.

File: tests/deptest_routes.py [hunk 12]
```diff
@@ -0,0 +1,265 @@
+import time
+
+from tests.helpers import count_fund_applications
+from tests.helpers import expected_data_within_get_response
+from tests.helpers import post_data
+
+
+# TODO: Not possible to test all applications currently as unpredictable UID in default data
```
Comment (endamccormack) [inline tests/deptest_routes.py:8]: Have you created an extra story to address this if you're not going to address it in this PR?

Comment (endamccormack) [reply tests/deptest_routes.py:8]: I'd prefer if you removed these and left them on a branch

**Review by endamccormack — APPROVED** [review]
A couple of suggestions included but looks good

File: dockerfile (hunk 10 above)
Comment (harryyo) [reply dockerfile:11]: Yep was for local work shouldn't have committed

Comment (harryyo) [reply dockerfile:11]: Removed

File: tests/deptest_routes.py (hunk 12 above)
Comment (harryyo) [reply tests/deptest_routes.py:8]: These are deprecated tests - removing

File: api/namespace/applications/routes.py (hunk 11 above)
Comment (harryyo) [reply api/namespace/applications/routes.py:112]: The updated_section is json serialisable so rest-x should be able to handle it (as well as make_response)
